Add ability to prompt user before applying fix

rubberduck203 opened the pull request:
Use Case
---
We have some fixes that we don't want to apply without user approval because there will be data loss, e.g. `rails db:reset`.

Solution
---
Extend the current `ScopeDoctorGroup.spec.fix` element to allow users to specify that a fix shouldn't be applied without user approval.

```yaml
      fix:
        prompt:
          text: |-
            This may destroy some data.
            Do you wish to continue?
          # this is an optional field
          helpText: >-
            Some additional context about why this needs approval
            and what it's actually doing
```

Current implementation achieves this by adding a nullable `prompt` property to the schema.
When it's unset, we get the current behavior, so this is backward compatible.

Example can be seen by checking out this branch and running

```sh
cargo run -- doctor run --working-dir examples --only prompt
```

https://github.com/user-attachments/assets/42509a95-b1b8-4579-a6ff-646787ed91bb

This PR depends on #183 which fixed some issues with the summary calculation.

rubberduck203 commented on scope/src/doctor/runner.rs:
```diff
@@ -230,6 +246,17 @@ where
+        inquire::Confirm::new(prompt_text)
+            .with_default(false)
+            //FIXME: do we want to allow a help message here?
+            // .with_help_message("additional text explaining why this needs approval")
```
Need to decide on this now because changing it would be a breaking change.

rubberduck203 replied:
I decided to just go ahead and provide the mechanism to display an additional help message.

rubberduck203 commented on scope/src/doctor/commands/run.rs:
```diff
@@ -80,13 +80,16 @@ pub async fn doctor_run(found_config: &FoundConfig, args: &DoctorRunArgs) -> Res
-                )
-                .prompt()
-                .unwrap_or(false)
+            // FIXME: CMAC make sure we actually need this suspend when we're done
```
This FIXME needs addressed

rubberduck203 commented on scope/src/doctor/check.rs:
```diff
@@ -289,17 +304,38 @@ impl DefaultDoctorActionRun {
+
+        for command in commands {
+            let report = self.run_single_fix(command).await?;
+            highest_exit_code = max(highest_exit_code, report.exit_code.unwrap_or(NO_COMMANDS_EXIT_CODE));
```
I think maybe the `-1` exit code was being overloaded.  
Should we add another exit code?

rubberduck203 commented on scope/src/doctor/check.rs:
```diff
@@ -179,10 +185,10 @@ impl DoctorActionRun for DefaultDoctorActionRun {
 
         match fix_result {
-            i32::MIN..=-1 => {
+            i32::MIN..=-3 | NO_COMMANDS_EXIT_CODE => {
```
I feel like the `i32::MIN..=-3` range should not return an `Ok(NoFixProvided)` as exit codes in that range are unexpected.

rubberduck203 commented on scope/src/doctor/check.rs:
```diff
@@ -289,17 +310,41 @@ impl DefaultDoctorActionRun {
+            let report = self.run_single_fix(command).await?;
+            highest_exit_code = max(
+                highest_exit_code,
+                report.exit_code.unwrap_or(NO_COMMANDS_EXIT_CODE),
```
I feel like the `-1` exit code was being used for two separate exit reasons.  
Would be happy to follow up with another PR to disambiguate exit codes or slip a commit into this one if you agree this should have a different exit code.

noizwaves commented on examples/.scope/doctor-group-prompt.yaml:
```diff
@@ -0,0 +1,28 @@
+            This may destroy some data.
+            Do you wish to continue?
+          # this is an optional field
+          helpText: >-
```
In other parts of the schema, `helpText` refers to "humanized instructions to fix a problem", but that's quite different to this. What about a different name, like `extraContext`?

rubberduck203 replied:
I had the same concern. I like your suggestion. Will update the code.

rubberduck203 replied:
This is done.

noizwaves commented on scope/src/doctor/runner.rs:
```diff
@@ -603,61 +644,173 @@ mod tests {
-                "group_1".to_string(),
-                "group_2".to_string(),
-                "group_3".to_string(),
+                "fails".to_string(),
+                "skipped_1".to_string(),
+                "skipped_2".to_string(),
```
If possible, please split these test refactorings out into a separate PR. That will help reduce the size of this behavior changing PR, whilst also improving code quality (and helping speed up PR approval).

rubberduck203 replied:
Not a problem. I thought this diff would go away once #183 was merged. Will look into what’s going on here and either update this PR or submit one with just these refactors.

rubberduck203 replied:
A rebase fixed this up.

rubberduck203 commented:
@noizwaves I addressed your feedback and cleaned up history. 
Each commit passes a `cargo test && cargo fmt -- --check && cargo clippy -- -D warnings` test.

I think it's ready for a real review now, so I moved it out of draft status.

rubberduck203 commented:
I just realized that `scope doctor run` exiting with a successful status code is problematic when utilizing it in other scripts. 
I need to update this to ensure it exits with failure when the user says "no".

rubberduck203 commented:
> I just realized that `scope doctor run` exiting with a successful status code is problematic when utilizing it in other scripts. I need to update this to ensure it exits with failure when the user says "no".

98a6f75 addresses this.
